test/tool/net/sha_test.lua

```diff
@@ -0,0 +1,89 @@
+-- Copyright 2022 Justine Alexandra Roberts Tunney
+--
+-- Permission to use, copy, modify, and/or distribute this software for
+-- any purpose with or without fee is hereby granted, provided that the
+-- above copyright notice and this permission notice appear in all copies.
+--
+-- THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL
+-- WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED
+-- WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE
+-- AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL
+-- DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR
+-- PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER
+-- TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
+-- PERFORMANCE OF THIS SOFTWARE.
+
+-- RFC 6234 - https://datatracker.ietf.org/doc/html/rfc6234#section-8.5
+--    This document replaces RFC 4634, fixing errata and adding code for an
+--    HMAC-based extract-and-expand Key Derivation Function, HKDF (RFC
+--    5869).  As with RFC 4634, code to perform SHA-based Hashed Message
+--    Authentication Codes (HMACs) is also included.
+
+-- SHA-1
+assert(Sha1("abc") == "\xa9\x99\x3e\x36\x47\x06\x81\x6a\xba\x3e\x25\x71\x78\x50\xc2\x6c\x9c\xd0\xd8\x9d")
+assert(
+    Sha1("abcdbcdecdefdefgefghfghighijhijkijkljklmklmnlmnomnopnopq") ==
+    "\x84\x98\x3e\x44\x1c\x3b\xd2\x6e\xba\xae\x4a\xa1\xf9\x51\x29\xe5\xe5\x46\x70\xf1"
+)
+assert(
+    Sha1(string.rep("a", 1000000)) == "\x34\xaa\x97\x3c\xd4\xc4\xda\xa4\xf6\x1e\xeb\x2b\xdb\xad\x27\x31\x65\x34\x01\x6f"
+)
+assert(
+    Sha1(string.rep("0123456701234567012345670123456701234567012345670123456701234567", 10)) ==
+    "\xde\xa3\x56\xa2\xcd\xdd\x90\xc7\xa7\xec\xed\xc5\xeb\xb5\x63\x93\x4f\x46\x04\x52"
+)
+-- TODO - dunno how the extrabits and numberExtrabits are handled yet (..., 0x98, 5, ...)
+-- i.e. - { "", 0, 0, 0x98, 5, "29826B003B906E660EFF4027CE98AF3531AC75BA" }
+-- assert(Sha1("") ==
+--         "\x29\x82\x6b\x00\x3b\x90\x6e\x66\x0e\xff\x40\x27\xce\x98\xaf\x35\x31\xac\x75\xba")
```

Comment:
Could use some help understanding these extrabits and numberExtrabits, and how they are used / effect the test case